Consensus sieve has error on both LoZ and X86 platforms -- one peer cannot get block

## Description

<!--- Describe your issue or user story in detail. -->

Get the latest hyper ledger code (commit 66ca505deb06f4df024f0540d18bfdecb04584fa )
Run with busywork and get error from peers. After tested with @bcbrock , he confirmed that this is bug with  sieve on LoZ platform.
## Describe How to Reproduce

<!--- If an issue, provide sufficient context and steps to reproduce the issue -->

1) Get the latest hyper ledger code;
2) Build peer ;
3) run "make stress2s" from folder "$GPATHsrc/github.com/hyperledger/fabric/tools/busywork/counters"

Error msg:

```
 ...
 driver:     Client 0 is subprocess 49966
 driver: Waiting (indefinitely) for subprocesses to complete
 fabricLogger:ERROR: Block 2206 JSON response does not parse
 fabricLogger:ERROR: HTTP 'ncode' = '200'
 fabricLogger:ERROR: Full HTTP token dump below; Aborting
 fabricLogger:ERROR:     sock socka8d15a80
 fabricLogger:ERROR:     http HTTP/1.1 200 OK
 fabricLogger:ERROR:     querylength 0
 fabricLogger:ERROR:     -queryblocksize 8192
 fabricLogger:ERROR:     state body
 fabricLogger:ERROR:     charset iso8859-1
 fabricLogger:ERROR:     type application/json
 fabricLogger:ERROR:     -binary false
 fabricLogger:ERROR:     url http://0.0.0.0:48106/chain/blocks/2206
 fabricLogger:ERROR:     -keepalive 0
 fabricLogger:ERROR:     -timeout 0
 fabricLogger:ERROR:     socketinfo 0.0.0.0:48106
 fabricLogger:ERROR:     queryoffset 0
 fabricLogger:ERROR:     binary 1
 fabricLogger:ERROR:     -strict 1
 fabricLogger:ERROR:     totalsize 5
 fabricLogger:ERROR:     -validate 0
 fabricLogger:ERROR:     -protocol 1.1
 fabricLogger:ERROR:     connection close
 fabricLogger:ERROR:     -queryprogress
 fabricLogger:ERROR:     -headers
 fabricLogger:ERROR:     -blocksize 8192
 fabricLogger:ERROR:     coding
 fabricLogger:ERROR:     status ok
 fabricLogger:ERROR:     body null

 fabricLogger:ERROR:     currentsize 5
```

Then use `url http://localhost:<peer REST port>/chain/blocks/3` to find it returns "null" from vp3 only,
All 4 vps are running.

Here is error
[vps_log.zip](https://github.com/hyperledger/fabric/files/274937/vps_log.zip)
 log from vps

Tested on a x-86 host, got the similar error:

```
...   
driver: Waiting up to 60s for array creation to be registered
driver: Spawning clients:
driver:     Client 0 is subprocess 5591
driver: Waiting (indefinitely) for subprocesses to complete
fabricLogger:ERROR: REST API call to 0.0.0.0:36008 for block 1574 failed with 0 retries
fabricLogger:ERROR: HTTP 'ncode' = ''
fabricLogger:ERROR: Full HTTP token dump below; Aborting
fabricLogger:ERROR:     sock sock9600ee0
fabricLogger:ERROR:     http
fabricLogger:ERROR:     querylength 0
fabricLogger:ERROR:     -queryblocksize 8192
fabricLogger:ERROR:     state connecting
fabricLogger:ERROR:     charset iso8859-1
fabricLogger:ERROR:     type text/html
fabricLogger:ERROR:     -binary false
fabricLogger:ERROR:     url http://0.0.0.0:36008/chain/blocks/1574
fabricLogger:ERROR:     -keepalive 0
fabricLogger:ERROR:     -timeout 0
fabricLogger:ERROR:     socketinfo 0.0.0.0:36008
fabricLogger:ERROR:     queryoffset 0
fabricLogger:ERROR:     binary 0
fabricLogger:ERROR:     -strict 1
fabricLogger:ERROR:     totalsize 0
fabricLogger:ERROR:     -validate 0
fabricLogger:ERROR:     -protocol 1.1
fabricLogger:ERROR:     connection close
fabricLogger:ERROR:     -queryprogress
fabricLogger:ERROR:     -headers
fabricLogger:ERROR:     -blocksize 8192
fabricLogger:ERROR:     coding
fabricLogger:ERROR:     status ok
fabricLogger:ERROR:     body
fabricLogger:ERROR:     currentsize 0
fabricLogger:ERROR:     meta
fabricLogger:ERROR:     -type application/x-www-form-urlencoded
Makefile:92: recipe for target 'stress2s' failed
```

Here is full vp log 
[sieve_error_log_x86.zip](https://github.com/hyperledger/fabric/files/275339/sieve_error_log_x86.zip)

@samding01 Peer 2 crashed in your logs:

```
[33m15:29:28.824 [peer] beforeSyncBlocks -> WARN 251[0m Ignoring SyncBlocks message with correlationId = 3, blocks 1344 to 1344, as current correlationId = 4
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x9faa97]

goroutine 11 [running]:
github.com/hyperledger/fabric/consensus/obcpbft.(*pbftCore).execDoneSync(0xc820077ce0)
    /gsa/tlbgsa/home/s/a/samding/web/public/workplace/fab/src/github.com/hyperledger/fabric/consensus/obcpbft/pbft-core.go:902 +0x247
github.com/hyperledger/fabric/consensus/obcpbft.(*pbftCore).main(0xc820077ce0)
    /gsa/tlbgsa/home/s/a/samding/web/public/workplace/fab/src/github.com/hyperledger/fabric/consensus/obcpbft/pbft-core.go:299 +0x886
created by github.com/hyperledger/fabric/consensus/obcpbft.newPbftCore
    /gsa/tlbgsa/home/s/a/samding/web/public/workplace/fab/src/github.com/hyperledger/fabric/consensus/obcpbft/pbft-core.go:255 +0x167d
```

I reran just now and I also saw the same failure on the terminal, however I stupidly ran again w/o checking the VP logs. Now it is running fine for 5 times. So this is an intermittent error. You may want to change the title to remove that it is Z-specific, I think it is a cross-platform issue.

I caught it again on my system, same crash as Sam:

```
[33m15:20:49.438 [peer] beforeSyncBlocks -> WARN 302[0m Ignoring SyncBlocks message with correlationId = 4, blocks 4964 to 4964, as current correlationId = 6
[33m15:20:49.438 [peer] beforeSyncBlocks -> WARN 303[0m Ignoring SyncBlocks message with correlationId = 4, blocks 4963 to 4963, as current correlationId = 6
[33m15:20:49.438 [peer] beforeSyncBlocks -> WARN 304[0m Ignoring SyncBlocks message with correlationId = 4, blocks 4962 to 4962, as current correlationId = 6
[33m15:20:49.438 [peer] beforeSyncBlocks -> WARN 305[0m Ignoring SyncBlocks message with correlationId = 4, blocks 4961 to 4961, as current correlationId = 6
[33m15:20:50.332 [consensus/obcpbft] weakCheckpointSetOutOfRange -> WARN 306[0m Replica 3 is out of date, f+1 nodes agree checkpoint with seqNo 5410 exists but our high water mark is 5400
[33m15:20:50.385 [consensus/obcpbft] weakCheckpointSetOutOfRange -> WARN 307[0m Replica 3 is out of date, f+1 nodes agree checkpoint with seqNo 5460 exists but our high water mark is 5450
[33m15:20:50.425 [consensus/obcpbft] weakCheckpointSetOutOfRange -> WARN 308[0m Replica 3 is out of date, f+1 nodes agree checkpoint with seqNo 5510 exists but our high water mark is 5500
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x9fb1f7]

goroutine 23 [running]:
panic(0xd1a140, 0xc820010150)
    /usr/src/go/src/runtime/panic.go:464 +0x3e6
github.com/hyperledger/fabric/consensus/obcpbft.(*pbftCore).execDoneSync(0xc8202429a0)
    /gsa/ausgsa/home/b/c/bcbrock/hyperledger/gopath/src/github.com/hyperledger/fabric/consensus/obcpbft/pbft-core.go:902 +0x247
github.com/hyperledger/fabric/consensus/obcpbft.(*pbftCore).main(0xc8202429a0)
    /gsa/ausgsa/home/b/c/bcbrock/hyperledger/gopath/src/github.com/hyperledger/fabric/consensus/obcpbft/pbft-core.go:299 +0x886
created by github.com/hyperledger/fabric/consensus/obcpbft.newPbftCore
    /gsa/ausgsa/home/b/c/bcbrock/hyperledger/gopath/src/github.com/hyperledger/fabric/consensus/obcpbft/pbft-core.go:255 +0x167d
```

--- a/consensus/obcpbft/pbft-core.go
+++ b/consensus/obcpbft/pbft-core.go
@@ -894,7 +894,9 @@ func (instance *pbftCore) execDoneSync() {
 		}
 
 	} else {
-		logger.Debug("Replica %d had execDoneSync called, but there is no current execution", instance.id)
+		// XXX This masks a bug, this should not be called when currentExec is nil
+		logger.Warning("Replica %d had execDoneSync called, flagging ourselves as out of date", instance.id)
+		instance.skipInProgress = true
 	}
 	instance.currentExec = nil
 

--- a/consensus/obcpbft/pbft-core_test.go
+++ b/consensus/obcpbft/pbft-core_test.go
@@ -1291,3 +1291,8 @@ func TestReplicaPersistDelete(t *testing.T) {
 		t.Error("expected no persisted entry")
 	}
 }
+
+func TestNilCurrentExec(t *testing.T) {
+	p := newPbftCore(1, loadConfig(), &omniProto{})
+	p.execDoneSync() // Per issue 1538, this would cause a Nil pointer dereference
+}

